x-pack/plugins/licensing/public/plugin.test.ts

```diff
@@ -115,7 +115,8 @@ describe('licensing plugin', () => {
             refresh();
           } else if (i === 2) {
             expect(value.type).toBe('gold');
-            refresh();
+            // if we just call refresh() immediately here, the exhaustMap de-dupes the call
+            Promise.resolve().then(() => refresh());
```

Comment:
I think there might be a bug in RxJs. It's not clear to me why this doesn't work as expected https://stackblitz.com/edit/exhaustmap-1 but https://stackblitz.com/edit/exhaustmap-2 and https://stackblitz.com/edit/exhaustmap-3 do...

Reply:
Not sure if a bug or an implementation subtlety when the consumer is synchronous... @rudolf do you have an opinion?

Anyway, if it's working with the `Promise.resolve` trick, the comment is probably good enough.

May want to change the `Promise.resolve` with a `setImmediate` though.

Reply:
It does appear to be related to the synchronous subscriber. `subscribeToPromise` is first calling `subscriber.next()` before immediately calling `subscriber.complete()` [here](https://github.com/ReactiveX/rxjs/blob/5c68e208512dd0f3ba1aafbf3e592592ec13db9b/src/internal/util/subscribeToPromise.ts#L8-L9). If the subscription to `exhaustMap` is synchronous, it doesn't have the chance to mark the subscriber as "complete" before we emit another value onto the subject.

In my simplistic mental model, it shouldn't matter that the subscription is synchronous. Is that not the case in your alls experience?

With regard to `Promise.resolve` vs `setImmediate`, either will work. After giving this more thought, I think we should use `process.nextTick`. Borrowing a diagram of the event-loop from https://blog.insiderattack.net/timers-immediates-and-process-nexttick-nodejs-event-loop-part-2-2c53fd511bb we don't necessarily need to wait until the event-loop is in the "Immediate Queue" phase, we just need some mechanism of allowing `subscriber.complete()` to be called. The `Promise.resolve` approach will run this operation as part of the "Microtasks", which works, but is somewhat confusing. Where-as `process.nextTick` will perform a similar function but more explicitly. `setImmediate` will cause this to be run during a different phase of the event-loop, which works, but we don't necessarily need this to be run in a specific phase:

![1_2yXbhvpf1kj5YT-m_fXgEQ](https://user-images.githubusercontent.com/627123/93486571-37fa2500-f8b9-11ea-92d0-835e64eb0c6b.png)

Reply:
> It does appear to be related to the synchronous subscriber. subscribeToPromise is first calling subscriber.next() before immediately calling subscriber.complete() here. If the subscription to exhaustMap is synchronous, it doesn't have the chance to mark the subscriber as "complete" before we emit another value onto the subject.

Might be a bug in rxjs implementation then. 

> In my simplistic mental model, it shouldn't matter that the subscription is synchronous

Yea, it shoudn't, theoretically. However in my experience, it does sometimes 😄. TBH I always wondered why they choose to handle subscriptions / emissions synchronously. This always leads to some borderline cases.

> With regard to Promise.resolve vs setImmediate, either will work. After giving this more thought, I think we should use process.nextTick

Fine with me. TBH, for such a 'small' delayed execution, it probably doesn't matter much. But I agree with you on paper

Reply:
Cool, I'm going to proceed with this approach and open an issue in the RxJs repo about synchronous subscriptions.

Reply:
https://github.com/ReactiveX/rxjs/issues/5658#issuecomment-694888866